prism-agent/service/server/src/main/scala/io/iohk/atala/iam/authentication/SecurityLogic.scala

```diff
@@ -75,4 +78,32 @@ object SecurityLogic {
         case Left(entity)  => authorizeWalletAdmin(entity)(EntityAuthorizer).map(entity -> _)
         case Right(entity) => authorizeWalletAdmin(entity)(authorizer).map(entity -> _)
       }
+
+  def authorizeRole[E <: BaseEntity](credentials: Credentials, others: Credentials*)(
+      authenticator: Authenticator[E],
+  )(permittedRole: EntityRole): IO[ErrorResponse, BaseEntity] = {
+    authenticate[E](credentials, others: _*)(authenticator)
+      .flatMap { ee =>
+        val entity = ee.fold(identity, identity)
+        for {
+          role <-
+            ZIO
+              .fromEither(entity.role)
+              .mapError(msg =>
+                AuthenticationError.UnexpectedError(s"Unable to retrieve entity role for entity id ${entity.id}. $msg")
+              )
+              .mapError(AuthenticationError.toErrorResponse)
+          _ <- ZIO
+            .fail(AuthenticationError.InvalidRole(s"$role role is not permitted. Expected $permittedRole role."))
+            .when(role != permittedRole)
+            .mapError(AuthenticationError.toErrorResponse)
```

Comment:
We need to keep the Authentication and Authorization layers isolated.
If the role is not permitted - it should be an Authorization error.
Authentication is about resolving the Entity by Credentials
Authorization is about checking that Entity has particular permissions.

Reply:
I'll raise a tech-debt for this and tackle it as part of a bigger refactor. As you can see, we reuse `AuthenticationError` for authentication and authorization :smile: